freqtrade/exchange/exchange.py

```diff
@@ -1704,6 +1746,141 @@ def set_margin_mode(self, pair: str, collateral: Collateral, params: dict = {}):
+                since=since,
+                params={
+                    'price': self._ft_has["mark_ohlcv_price"]
+                }
+            )
+            history = {}
+            for candle in candles:
+                d = datetime.fromtimestamp(int(candle[0] / 1000), timezone.utc)
+                # Round down to the nearest hour, in case of a delayed timestamp
+                # The millisecond timestamps can be delayed ~20ms
+                time = datetime(
+                    d.year,
+                    d.month,
+                    d.day,
+                    d.hour,
+                    tzinfo=timezone.utc
+                ).timestamp() * 1000
+                opening_mark_price = candle[1]
+                history[time] = opening_mark_price
+            return history
+        except ccxt.NotSupported as e:
+            raise OperationalException(
+                f'Exchange {self._api.name} does not support fetching historical '
+                f'mark price candle (OHLCV) data. Message: {e}') from e
+        except ccxt.DDoSProtection as e:
+            raise DDosProtection(e) from e
+        except (ccxt.NetworkError, ccxt.ExchangeError) as e:
+            raise TemporaryError(f'Could not fetch historical mark price candle (OHLCV) data '
+                                 f'for pair {pair} due to {e.__class__.__name__}. '
+                                 f'Message: {e}') from e
+        except ccxt.BaseError as e:
+            raise OperationalException(f'Could not fetch historical mark price candle (OHLCV) data '
+                                       f'for pair {pair}. Message: {e}') from e
+
+    def calculate_funding_fees(
+        self,
+        pair: str,
+        amount: float,
+        open_date: datetime,
+        close_date: Optional[datetime] = None
+    ) -> float:
+        """
+            calculates the sum of all funding fees that occurred for a pair during a futures trade
+            :param pair: The quote/base pair of the trade
+            :param amount: The quantity of the trade
+            :param open_date: The date and time that the trade started
+            :param close_date: The date and time that the trade ended
+        """
+
+        fees: float = 0
+        if not close_date:
+            close_date = datetime.now(timezone.utc)
+        open_timestamp = int(open_date.timestamp()) * 1000
+        # close_timestamp = int(close_date.timestamp()) * 1000
+        funding_rate_history = self.get_funding_rate_history(
+            pair,
+            open_timestamp
+        )
+        mark_price_history = self._get_mark_price_history(
+            pair,
+            open_timestamp
+        )
+        funding_fee_dates = self._get_funding_fee_dates(open_date, close_date)
+        for date in funding_fee_dates:
+            timestamp = int(date.timestamp()) * 1000
+            funding_rate = funding_rate_history[timestamp]
+            mark_price = mark_price_history[timestamp]
```

Comment:
this is bound to fail.
assuming an exchange downtime that spans for 5 hours - you will encounter 1 missing timestamp.

As you're not checking if timestamp is in the index (nor access it safely with `.get()`) - this will make freqtrade fail / crash until the downtime period is no longer relevant (it either ages out - or the trade is deleted).